---
title: "edx HDDA"
author: "Maarten van Schaik"
date: "March 30, 2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r, eval=FALSE}
library(devtools)
library(rafalib)
# install_github("genomicsclass/GSE5859Subset")
# install_bioc("genefilter")
# install_bioc("qvalue")
library(GSE5859Subset)
library(genefilter)
library(qvalue)
```

# Week 2

## FDR Exercises #1

Load the gene expression data. There are 24 samples, and for each sample there are 8793 genes. The samples in `sampleInfo` are in the same order as the columns in `geneExpression`.

```{r}
library(GSE5859Subset)
data(GSE5859Subset)
head(sampleInfo)
dim(sampleInfo)
dimnames(geneExpression)[[2]] == sampleInfo$filename
sampleInfo$group
```

We are interested in comparing gene expression between the two groups defined in the `sampleInfo` table.

Compute a p-value for each gene using the function rowttests from the `genefilter` package in Bioconductor. It is much faster than using a `t.test` many times, and gives similar results. We see that 1383 out of 8793 are 'significant' at p=0.05:

```{r}
library(genefilter)
?rowttests
# First manually using apply and t.test:
# pvals1 <- apply(geneExpression, 1, function(x) t.test(x~sampleInfo$group)$p.val)
# now using rowttests:
pvals2 <- rowttests(x=geneExpression, fac=factor(sampleInfo$group))
sum(pvals2$p.value<0.05)
```

## FDR Exercises #2

Apply the Bonferroni correction to the p-values obtained in question #1 to achieve a FWER of 0.05. How many genes are called significant under this procedure?

```{r}
p <- 0.05
n_test <- nrow(geneExpression)
sum(pvals2$p.value<p/n_test)
```

## FDR Exercises #3

Note that the FDR is a property of a list of features, not each specific feature. The q-value relates FDR to an individual feature. To define the q-value we order features we tested by p-value then compute the FDRs for a list with the most significant, the two most significant, the three most significant, etc... The FDR of the list with the, say, m most significant tests is defined as the q-value of the m-th most significant feature. In other words, the q-value of a feature, is the FDR of the biggest list that includes that gene.

In R, we can compute the q-value using the `p.adjust` function with the FDR option. Read the help file for `p.adjust` and then, for our gene expression dataset, compute how many genes achieve an FDR < 0.05.

```{r}
?p.adjust
fdrs <- p.adjust(pvals2$p.value, method="fdr")
sum(fdrs<0.05)
```

## FDR Exercises #4

Now use the `qvalue` function, in the Bioconductor `qvalue` package, to estimate q-values using the procedure described by *Storey*.

Using this estimate how many genes have q-values below 0.05?

```{r}
require(qvalue)
?qvalue
qvals <- qvalue(pvals2$p.value, fdr.level=0.05)
sum(qvals$significant)
# also see this nice summary:
summary(qvals)
```

## FDR Exercises #5

Read the help file for qvalue and report the estimated proportion of genes for which the null hypothesis is true: $\pi_0 = m_0 / m$

```{r}
# also see this nice summary:
summary(qvals)
```

## FDR Exercises #6

Note that we have the number of genes passing the q-value `<0.05` threshold is larger with the `qvalue` function than the `p.adjust` difference.

Why is this the case? Make a plot of the ratio of these two estimates to help answer the question.

## FDR Exercises #7

Note that this is an advanced question and that you can ask questions in the discussion forum.

Create a Monte Carlo Simulation in which you simulate measurements from 8,793 genes for 24 samples: 12 cases and 12 controls.

```{r}
n <- 24
m <- 8793
mat <- matrix(rnorm(n*m),m,n)
```

Now for 500 genes, there is a difference of 2 between cases and controls:

```{r}
delta <- 2
positives <- 500
mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
```

So the null hypothesis is true for 8793-500 genes. Using the notation from the videos m=8793, m0=8293 and m1=500

Set the seed at 1, `set.seed(1)` and run this experiment 1,000 times with a Monte Carlo simulation. For each instance compute p-values using a t-test (using `rowttests` in the `genefilter` package) and create three lists of genes using:

1. Bonferroni correction to achieve an FWER of 0.05,
2. `p-adjust` estimates of FDR to achieve an FDR of 0.05, and
3. `qvalue` estimates of FDR to to achieve an FDR of 0.05.

For each of these three lists compute the number of false positives in the list and the number of false negatives: genes not in the list that should have been because the null hypothesis is not true (we added 2 to the controls to create the cases).

What is the false positive rate (false positives divided by m0) if we use Bonferroni?

```{r}

```

## FDR Exercises #8

From the same Monte Carlo simulation as in the question above, what is the false negative rate if we use Bonferroni?

## FDR Exercises #9

From the same Monte Carlo simulation as in question #7, what is the *false positive* rate if we use q-values from `p.adjust`?

## FDR Exercises #10

From the same Monte Carlo simulation as in question #7, what is the *false negative* rate if we use q-values from `p.adjust`?

## FDR Exercises #11

From the same Monte Carlo simulation as in question #7, what is the *false positive rate* if we use q-values from `qvalue` function?

## FDR Exercises #12

From the same Monte Carlo simulation as in question #7, what is the *false negative rate* if we use q-values from `qvalue` function?
